### DISCUSSION TITLE
Plugging in a new linear solver

### DISCUSSION THREAD

**Discussion by dpo** [coin-or/Ipopt#618] [General]
Hi there.

IPOPT already supports several linear solvers (MUMPS, Pardiso, HSL, ...). Say I have a new linear solver and I would like to hook it up to IPOPT. Is there a mechanism for me to implement a subclass of `AugSystemSolver` (say) and ask IPOPT to use my linear solver? Do I have to build the interface to my solver as part of IPOPT in order to do so, or is there a way to keep both codes completely separate?

Thank you.

**Comment by svigerske** [comment]
To keep things separate, I think the idea is that your use the `OptimizeNLP()` function that takes an `AlgorithmBuilder` object (https://coin-or.github.io/Ipopt/classIpopt_1_1IpoptApplication.html#aa23ec7c9232307fbdc37a2121acdc373).

The [`AlgorithmBuilder`](https://coin-or.github.io/Ipopt/classIpopt_1_1AlgorithmBuilder.html) object is easy to construct, but you should use the constructor that takes an [`AugSystemSolver`](https://coin-or.github.io/Ipopt/classIpopt_1_1AugSystemSolver.html) as argument. This object should then be your interface to a linear solver. In addition, you will need to set option `linear_solver` to `custom`.

PS: Looks like you shouldn't implement an `AugSystemSolver` directly. But just follow the code from `AlgorithmBuilder` on how it constructs an `AugSystemSolver` from a [`SparseSymLinearSolverInterface`](https://coin-or.github.io/Ipopt/classIpopt_1_1SparseSymLinearSolverInterface.html). The latter is what all linear solver interfaces implement.

**Reply by dpo** [reply]
Many thanks @svigerske! I'll let you know how it goes.

**Reply by dpo** [reply]
@svigerske I've just come back to this. Thank you for your guidance. I managed to get things to work with a couple of hiccups:

1. some IPOPT headers were missing because, for some reason, they are not installed. I installed them by hand but it seems necessary to install them by default (in fact, why not install all headers?);
2. my solver requires CSR format. However, when I asked IPOPT to provide the augmented system in `CSR_Format_0_Offset` format, it provided the *upper* triangle, not the *lower* triangle, as documented. Either this is an error in the documentation (but I don't think so because the lower triangle is returned for systems in triplet format), there's a mixup in the COO → CSR conversion, or you meant "CSC" instead of "CSR".

I opened two issues about the above: https://github.com/coin-or/Ipopt/issues/637 and https://github.com/coin-or/Ipopt/issues/638.

**Reply by svigerske** [reply]
Installing additional headers is no problem. I think the original idea behind not installing all headers was that only those that belong to the Ipopt "API" should be necessary. But then advanced users requested the installation of more headers to access more of Ipopt's internal datastructure and more and more headers were installed. Hmm, I'm just going to follow this and install a few more headers on request.

If we were to change the matrix conversion routines, existing interfaces to linear solvers would need to be changed. So I would rather go with small changes to the documentation.

**Reply by dpo** [reply]
@svigerske Is there a hook for IPOPT to display linear-solver-specific statistics after the solve has ended?

**Reply by svigerske** [reply]
Setting option `print_timing_statistics` to `yes` will print more details about time spend in various components (incl linear solver) to the default journal.

This is the information stored in a [TimingStatistics](https://coin-or.github.io/Ipopt/classIpopt_1_1TimingStatistics.html) object.
The object should be accessible via the [TimingStats](https://coin-or.github.io/Ipopt/classIpopt_1_1IpoptData.html#a53fde5ffec9d56518fb90db19d39aac7) function from the IpoptData object.
The IpoptData object you can [get](https://coin-or.github.io/Ipopt/classIpopt_1_1IpoptApplication.html#a4d0682d8b1ec5751b8727bb0573956b3) from the IpoptApplication object.

Other than that, I don't think that there are more relevant statistics collected.

**Reply by dpo** [reply]
Thank you for all your guidance so far. I have a few follow-up questions:
1. the SpRal and Mumps interfaces have a `GetName()` method, but the other solvers don't. I couldn't find a mechanism to get IPOPT to query the name of a custom linear solver (at the beginning of the solve, Ipopt prints an empty string as the solver name). Is there one?
2. Is there a way to find out what primal and dual regularization values IPOPT used to perturb the diagonal of the augmented matrix (without manually inspecting the nonzeros)?

Thanks!

**Reply by svigerske** [reply]
1. That the string was empty was a small bug. It should now print "custom". And you can now pass in a name as additional argument for the `AlgorithmBuilder` constructor, which should then be printed instead of "custom". Is that sufficient, or do you need to be able to query the name as well via some function?

2. I think you can get this via [IpoptData::getPDPert()](https://coin-or.github.io/Ipopt/classIpopt_1_1IpoptData.html#aee3d699d4f633bb6d5ea6a67d7061bde). This should be the values that you see in the log (print_level 6) in the lines `Perturbation parameters: delta_x=... delta_s=... ...`
